### ISSUE TITLE
A none micro benchmark.

### ISSUE THREAD

**Issue by bd82** [mindoro/mindoro#4]
It would be interesting to see a none micro benchmark.
Meaning something that parses a large file.

I've tried modifying the existing benchmark to parse mindoro.js from the dist folder
but that results in some syntax error.

I've then tried some other files from the node modules directory as inputs (acorn.js/benchmark.js/...)
But instead of a syntax error no numerical result was posted.

In both cases the other parsers in the benchmark did produce meaningful results.

**Comment by KFlash** [comment]
Hi. As mentioned in #1 I'm still implementing code. The errors occurs because at this stage the code are uncomplete. However. I ran a benchmark myself 2 weeks ago with the complete source code. In all tests Mindoro was double or tripple as fast.

What you can do until the entire code are published. Is too make crazy long benchmarks with various combos and you will get an idea how fast Mindoro is.

**Comment by bd82** [comment]
Thanks @KFlash I will re-check in the future when this project more complete.

When you say passes TC39 you mean passes these conformance tests?
https://github.com/tc39/test262

### Regarding performance:
* What makes midoro (potentially when completed...) so much faster than tools such as Acorn and Esprima? Are there some general principles applied or just a lot of micro optimizations?

* Do you know if the scanner portions of midoro are also much faster than scanner portions of esprima/acorn or does most of the advantage comes from the parser parts?

The reason I'm asking is that I am trying to build an ECMA5 parser
myself built on top of a parsing library I've authored: 
https://github.com/SAP/chevrotain

and because ECMAScript cannot (easily) have a separate Lexing and Parsing phases
due to the RegExp and division ambiguity I am using an hand built lexer
based on existing JS parsers. So I should probably pick the fastest lexer to base my work on 😄

**Comment by KFlash** [comment]
@bd82 Yes, you are correct. The main goal with this project is to pass and faill all tests according to [TC39](https://github.com/tc39/test262). That is the main reason why things may go slow. I have to write tons of tests between everything I do. And I'm not pushing anything that doesn't pass unlike other parsers.

Regarding performance. The performance it shows now is the same performance it will have when it's completed or probably faster. I'm going to optimize it. My main goal is to be 4 - 6 times faster than other existing parsers. I'm only 3 -4 times faster ATM. 

The trick to get this kind of performance is basically use of bitmasks and the way I structured the whole 
thing mixed with some low-level knowledge from `Assembly` and `C`.

I'm aware of the regular expression issue. This will also be taken care of in the end. The source code for it exist, but not pushed yet.

**Comment by KFlash** [comment]
@bd82 I can also add that everything that have reached stage 3 will be added too - or have been added. This implementation such as dynamic import and rest properties can be enabled with the next option. Like this: `parseScript('function f({ x, y, ...z }) {}', { next: true })`. `Async iteration` are half way supported. It is supported for functions, and will be supported for `for in`. It also works for classes and object, but not fully implemented. Depends if I get time

**Comment by bd82** [comment]
> The trick to get this kind of performance is basically use of bitmasks and the way I structured the whole
thing mixed with some low-level knowledge from Assembly and C.

I will take a deeper look at your code, perhaps there are some patterns I could re-use in a generic parser library. I always thought libraries such as Esprima and Acorn are close to the max possible performance 😄 

> Within 7 - 8 days I guess this parser should be up and running.

Great I may try to extract lexer logic from mindoro for use in my ECMAScript 5 + Chevrotain POC.

**Comment by KFlash** [comment]
Note that all my main development are done in the dev branch. And I'm very soon ready to push something that will make this parser usable as long as you don't need arrows or templates!

**Comment by KFlash** [comment]
@bd82 Within 7 - 8 a complete parser will be pushed to Master branch. Everything are now working, and normal benchmarks can be run on my computer.

**Comment by bd82** [comment]
Great. 👍

**Comment by bd82** [comment]
I think i will try reusing the code from the dev bench and update it once you have a more final version.

I am only interested in the Lexer portions and I don't mind a few bugs and missing if it lets me POC my ideas 😄

**Comment by KFlash** [comment]
Note you need to track escaped identifiers and all that. Plus there exist
bugs with column values.

**Comment by KFlash** [comment]
BTW. There exist an location tracking issue ATM in the current scanner
code. The column values isn't always correct. And the identifier code isn't
completed. As you know Mindoro are down for an entire refactoring, and my
team of engineers are working on the entire parser code now. After this
changes have been pushed the rest of the code base will be fixed rapidly.
Included the known scanner issues.

**Comment by bd82** [comment]
Thanks, but I don't need location tracking or even advanced identifiers for POC purposes...
Only to be able to scan some large input file (for example lodash.js/jquery.js).
I can even **manually** fix some issues in the selected large input file to get it to scan
successfully (for example removing escaped identifiers).

If/when I will consider making the POC more productive I will have to go back and update the scanner code.

>  and my team of engineers are working on the entire parser code now

Eh? what team? I see only one contributor? 😕

**Comment by KFlash** [comment]
Ok. I understand. Team? Only me on that repo. Offline we are a team
working. Mostly over Skype. People I worked with before in my old job.

**Comment by bd82** [comment]
Alright.

I've experimented with extracting the lexing logic away(used master branch) from Mindoro.
What I've found out is that it is only slightly faster than Acorn's **tokenize**
API, while doing less work (Acorn creates more substantial token objects).

That leads me to some conclusions.
* The big performance advantage of Mindoro is probably at the parsing (not lexing parts)
* For Acorn most of the cpu time is spent on parsing parts.
   - verified by playing around with Acorn's bench suite. I got X4-5 more OPs/sec when only
      tokenizing versus full flow.

What this means for me is:
1. I should base my POC on Acorn as it is more mature and the speed benefit (lexing only) seems minor at best.
2. There is a greater potential for my POC (ECMAScript parser using Chevrotain parsing library)
    To sometime become productive and not just a "sample grammar" because even with the overhead of a library the performance may be competitive as Acorn/Esprima may still be far from max performance (in parsing part).
